torch/jit/_script.py

```diff
@@ -841,6 +849,93 @@ def call_prepare_scriptable_func(obj):
+        return obj
+    if isinstance(obj, ScriptFunction):
+        return obj
+
+    if isinstance(obj, torch.nn.Module):
+        obj = call_prepare_scriptable_func(obj)
+        return torch.jit._recursive.create_script_module(
+            obj, torch.jit._recursive.infer_methods_to_compile
+        )
+
+    qualified_name = _qualified_name(obj)
+
+    # If MonkeyType is installed, enable profile directed type annotation
+    # Check if example_inputs are defined and generate call traces
+    # for the method by running eager mode version of the method with
+    # the provide example inputs. This logs all the traces in type_trace_db
+    type_trace_db = JitTypeTraceStore()
+    if monkeytype_trace:
+        monkeytype_config = JitTypeTraceConfig(type_trace_db)
+        with monkeytype_trace(monkeytype_config):
+            for example_input in example_inputs:
+                obj(*example_input)
+
+    if inspect.isclass(obj):
+        # If this type is a `nn.Module` subclass, they probably meant to pass
+        # an instance instead of a Module
+        if issubclass(obj, torch.nn.Module):
+            raise RuntimeError(
+                "Type '{}' cannot be compiled since it inherits"
+                " from nn.Module,"
+                " pass an instance instead".format(obj)
+            )
+
+        # Enums are automatically usable in TorchScript, explicitly scripting
+        # is not necessary, but not harmful either.
+        if issubclass(obj, enum.Enum):
+            return obj
+
+        if not _is_new_style_class(obj):
+            raise RuntimeError(
+                "TorchScript classes must be new-style classes. "
+                "Please inherit from 'object'."
+            )
+        if len(obj.mro()) > 2:
+            raise RuntimeError(
+                "TorchScript classes does not support inheritance yet. "
+                "Please directly inherit from 'object'."
+            )
+        if _rcb is None:
+            _rcb = _jit_internal.createResolutionCallbackFromFrame(_frames_up + 1)
+        _compile_and_register_class(obj, _rcb, qualified_name)
+        return obj
+    else:
+        # this is a decorated fn, and we need to the underlying fn and its rcb
+        if hasattr(obj, "__script_if_tracing_wrapper"):
+            obj = obj.__original_fn
+            _rcb = _jit_internal.createResolutionCallbackFromClosure(obj)
+
+        _check_directly_compile_overloaded(obj)
+        maybe_already_compiled_fn = _try_get_jit_cached_function(obj)
+        if maybe_already_compiled_fn:
+            return maybe_already_compiled_fn
+        ast = get_jit_def(obj, obj.__name__)
+        if _rcb is None:
+            _rcb = _jit_internal.createResolutionCallbackFromClosure(obj)
+        fn = torch._C._jit_script_compile(
+            qualified_name, ast, _rcb, get_default_args(obj)
+        )
+        # Forward docstrings
+        fn.__doc__ = obj.__doc__
+        _set_jit_function_cache(obj, fn)
+        return fn
```

Comment:
Can we call `script` here instead of duplicating code? I guess this is because we want to run some sanity checks before calling MonkeyType, but it is probably less important during prototyping.

Reply:
Ok yeah, for now we can do that. Changed